Source/Particles/Resampling/VelocityCoincidenceThinning.H

```diff
@@ -106,10 +113,58 @@ public:
         }
     };
 
+    /**
+     * \brief Struct used to assign velocity space bin numbers to a given set
+     * of particles.
+    */
+    struct VelocityBinCalculator {
+
+        AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+        void operator() (const amrex::ParticleReal ux[], const amrex::ParticleReal uy[],
+                         const amrex::ParticleReal uz[], const unsigned int indices[],
+                         int bin_array[], int index_array[],
+                         const int cell_start, const int cell_stop) const
+        {
+            for (int i = cell_start; i < cell_stop; ++i)
+            {
+                int ii = -1, jj = -1, kk = -1;
+                if (velocity_grid_type == VelocityGridType::Spherical) {
+                    // get polar components of the velocity vector
+                    auto u_mag = std::sqrt(
+                        ux[indices[i]]*ux[indices[i]] +
+                        uy[indices[i]]*uy[indices[i]] +
+                        uz[indices[i]]*uz[indices[i]]
+                    );
+                    auto u_theta = std::atan2(uy[indices[i]], ux[indices[i]]) + MathConst::pi;
+                    auto u_phi = std::acos(uz[indices[i]]/u_mag);
+
+                    ii = static_cast<int>(u_theta / dutheta);
+                    jj = static_cast<int>(u_phi / duphi);
+                    kk = static_cast<int>(u_mag / dur);
+
+                }
+                else if (velocity_grid_type == VelocityGridType::Cartesian) {
+                    ii = static_cast<int>((ux[indices[i]] - ux_min) / dux);
+                    jj = static_cast<int>((uy[indices[i]] - uy_min) / duy);
+                    kk = static_cast<int>((uz[indices[i]] - uz_min) / duz);
```

Comment:
I was thinking of a way to avoid having to calculate the min and max. How does this look?
```
ii = static_cast<int>(std::fmod(ux[indices[i]] / dux, 1290.));
```
This will keep the indices spread out as long as the velocities don't fill out the space 1290*du. Note that 1290 is approximately the cube root of the max integer, the type of `bin_array`. Yes, I know this is a bit hacky but should work in most cases.

Reply:
That's a good point about being able to avoid the need to calculate max / min values by simply using the full range of integer values. The single cell index will be less straight forward to calculate when allowing negative `ii`, `jj` and `kk` values. But this can be circumvented by using
```
auto ux_min = 645 * dux;
ii = static_cast<int>((ux[indices[i]] - ux_min) / dux);
```
The question for me is whether avoiding the calculation of max / min values is worth sacrificing the algorithmic flexibility to have:

- velocity grids not centered at 0 m/s
- different number of cells in different directions - for example a species with a low temperature in `x` but a large temperature in `z` would naturally allow a small number of cells in `vx` (< 1290) and a large number of cells in `vz` (> 1290) if the same `du` values are used

Seeing as the min / max calculation is very likely not the limiting step in this scheme (it is GPU parallelized) I am inclined to favor the algorithmic flexibility over the minor performance improvement. Do you disagree?

Reply:
In my suggestion, negative values of indices should be valid, as long as they don't overlap with positive values from the neighboring cell, which would be the case if the particles don't fill the 1290*du space. The centering would in some sense be arbitrary (because of the modulus).
In the second case, if the same du was used, then the thinning in x would be poor since all of the particles would fall into a small number of bins so particles with widely varying ux would be combined. Also, if there's a case which needs more that 1290 bins, there are already a large number of particles per cell and there probably should be thinning happening earlier.
Though, if as you say the calculation of the min and max takes little time, then there is no need to avoid it. It is just a fun exercise to think about how it could be done.

Reply:
I agree it is fun to think about alternative and more clever ways of doing the combinatorics, so I would like to understand your strategy.
I tried to illustrate the problem I'm having with the negative indices in the following table (for a 2d grid to make the visualization easier):
![image](https://github.com/ECP-WarpX/WarpX/assets/40245517/280ed443-4ca3-4630-82df-553d47c79798)
The dashed lines represent the ux=0, uy=0 axes, the blue row gives the `ii` value for each cell and the orange row  the `jj` values. The integer in each cell is calculated from `idx = ii + jj * nx` where in this reduced case `nx = 3`. As you can see I get many cells with the same index, which is why I said the index value will be less straight forward to calculate. The top left and bottom right quadrants are fine but the others have duplicated cell indices. The all positive indices approach essentially shifts the origin so that only the top left quadrant is used. Am I missing something? Or do you have a different way of finding `idx` from `ii` and `jj`?

Reply:
I do agree that there can be overlap when using negative indices. But I'm suggesting using a large value for `nx`, with 1290 being a good value. There would only be overlaps if there are particles with for example velocities near 0 and near -1290*dux, which is unlikely. This is what I mean in my comment "negative values of indices should be valid, as long as they don't overlap with positive values from the neighboring cell, which would be the case if the particles don't fill the 1290*du space."